[Repository Management] Add quota display for repositories and attachments to projects

oliverguenther opened the pull request:
This PR adds the basic functionality for:
- Retrieving the storage size of a repository in pure ruby.
- Asynchronously retrieving this information from the physical repository and storing it into repositories.
- Displays cached storage requirements at
  - The admin projects table
  - The admin projects total information section
  - The project settings overview tab
### TODOS
- [x] Remove `.reposize` file in favor of database
- [x] Add route to repository WS to update storage information from external service.
- [x] Extract `acts_as_countable` as gem (https://github.com/finnlabs/acts_as_countable)
  - [x] countable specs
- [x] Project counting specs

Relevant work package: https://community.openproject.org/work_packages/18393

oliverguenther committed:
Add quota display using .reposize file.

This commits add the basic functionality for:

- Retrieving the storage size of a repository in pure ruby.
- Retrieving the latest mtime of a file in a repository
- Asynchronously getting and writing this information in a separated
file inside that repository (`.reposize`) with YAML for convenience.

TeatroIO commented:
I've prepared a stage to preview changes. <a href='http://oliverguenther-feature-repository-quota-display.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com' target='_blank'>Open stage</a> or <a href='http://oliverguenther-feature-repository-quota-display.opf-openproject-2af22bf12c7dfdad4ddf.ttrcloud.com/teatro.log' target='_blank'>view logs</a>.

ulferts commented:
**The good**
Appears to be working. 

The caching of the data collection is slick.

You got it working without having written any test.

**The bad**
There are no tests.

I dislike the implementation approach taken here. Storing the information in a file is working around the central storage facility we already have, the database. I see no benefit in doing this. Creating an additional table to hold the required information, or extending the `repositories` table would be the better solution:
- All the potential race conditions and locking is already handled for us when we use the database.
- The potential lack of write permissions on the directory (when the repository is included only to be shown via the UI) is also no problem. 
- Filtering and aggregation functions are easy to employ when using a db.  

**The ugly**
I do not like the way the size information is represented on the projects index page:

![image](https://cloud.githubusercontent.com/assets/617519/8931967/64895d40-353a-11e5-9d68-658256279432.png)

The layout suggests that the filtering to the top would influence the information below. It does influence the information in the table. However it does not change the data in the "INFORMATION" section. This section is IMO out of place. A quick fix would be to put the general information ("INFORMATION) to the top of the page and have the filtering/table below that. There probably is a better solution, still.

**The debatable**
We could extract `acts_as_countable` into it's own repository. Granted, it is not huge, but separating it before it becomes unwieldy could save effort later.

oliverguenther commented:
Yeah, this one should have definitely received a WIP tag from the start, my fault. I deliberately avoided specs at this stage due to timing constraints, as the whole functionality isn't really entirely specified and I wanted to have a discussion regarding it here and I wanted to avoid duplicating that effort.

I can see your point regarding robustness using the database for storing the reposize instead.
The benefit I foresaw in using an external file was that we can use post-receive hooks to update the value only when the physical repository change and it would be unfeasible to access the database in there.

The position of the information wasn't my decision, but there are a load of issues regarding the values. I  would prefer a count for the current set of projects on the current page (because I was asked to put projects pagination into 4.3  when humanly possible).

ulferts commented:
> I wanted to have a discussion

Well, that can be considered to be achieved already, I guess. :)

> The benefit I foresaw in using an external file was that we can use post-receive hooks to update the value only when the physical repository change and it would be unfeasible to access the database in there

Good point, hadn't thought of that one. Two ideas:
- we create a WS to trigger the data collection. The post commit hook could then call the WS to trigger the data collection. We are already using a WS for authentication/authorisation so it would be in line with our current approach. 
- We are using the existing authentication/authorisation WS endpoint to take note of the access to the repository. If it is a write access (would have to be transmitted by the Pearl-script), we trigger the data collection.

> The position of the information wasn't my decision, but there are a load of issues regarding the values. I would prefer a count for the current set of projects on the current page (because I was asked to put projects pagination into 4.3 when humanly possible).

Placing the information to to the top would reduce a lot of the issues as the filtering is then below the aggregated information and by that the user could deduce, that the two are independent. If pagination is introduced by subsequent PRs, good, change the position again.

oliverguenther committed:
Move storage information to database

This commit effectively adds a storage information column
to repositories.

It tests the functionality of the counting on subversion and git
repositories / adapters.

oliverguenther commented:
I would suggest to decouple this from authorization requests, as these may occur quite frequently in small time frames and we do not (yet) have the means to check whether the storage is already updating (i.e., reliably checking the status of a delayed job).

As I'm not sure yet how exactly the requirements on timely refreshing on repository storage will look like, we can probably make our lives easier to allow an optional cron job to update the information periodically.

In the current state, users will see outdated values once before the value is itself automatically updated through Delayed::Job. For normal users, this might be sufficient.
The value is now stored in the repositories table itself.

oliverguenther committed:
Extend SysController to allow WS to check / update storage

oliverguenther committed:
Added project counting specs

oliverguenther committed:
Cleanup

oliverguenther committed:
Extracted acts_as_countable into gem

oliverguenther commented:
@ulferts I've extracted `acts_as_countable` into its own gem. I've referred to it via branch for the time being, or should I release its initial version for this PR ?

Added specs for repository storage and project size counting. From my side, this should address all the remarks.

oliverguenther commented:
Rebased due to https://github.com/opf/openproject/pull/3292

NobodysNightmare commented on Gemfile:
```diff
@@ -45,6 +45,7 @@ gem 'warden-basic_auth', '~> 0.2.1'
 gem 'rails_autolink', '~> 1.1.6'
 gem "will_paginate", '~> 3.0'
 gem "acts_as_list", "~> 0.3.0"
+gem 'acts_as_countable', git: "https://github.com/finnlabs/acts_as_countable.git"
```
Should we specify any kind of version here?

I remember @ulferts saying that we should be more specific here...

I'd probably lock us to 0.1.x, i.e. only patch level releases. Even if we fully control the source gem.

edit: Since it is a git repo: Specify a tag or anything? Dunno...

oliverguenther replied:
I don't think I can set a specific version lock from the git repository, because it simply uses whatever is behind that refs gemspec.

We can probably fix it to the current commit (happens more often in our Gemfile) and update that explicitly?

NobodysNightmare replied:
I'd probably tag the current thing as `v1.0` or `v0.1` and use the tag? Or is that a bad idea?

NobodysNightmare commented:
I'd have expected to see some usage statistics on http://localhost:5000/projects/:id/repository/statistics, but this is only a usability/discoverability concern.

oliverguenther commented:
Right, this would make sense given its naming. We should probably add some more elaborate statistics there, not only the overall size.

NobodysNightmare commented:
Reading the requirements now:

> - Projects in the project table can be sorted numerically by used storage total.

Does not seem to work. Is there any update (on requirements/implementation) that I am missing?

NobodysNightmare commented:
It seems to me that the space required by my `:local` git repo is not taken into account in the information screens.

The Repo of my project "Chili con carne" reports 355000 `required_storage_bytes` (~355kB).

However my info screens just tell me:

![Required storage: 13 kB](https://cloud.githubusercontent.com/assets/453584/9063032/11c994bc-3ac6-11e5-8cb4-2ca6c73cf64c.png)
![Chili con carne 13 kB](https://cloud.githubusercontent.com/assets/453584/9063073/285a3d12-3ac6-11e5-97ea-a0d1470489af.png)

Hovering only displays allocated space for attachments, none for repositories.

oliverguenther commented:
> > Projects in the project table can be sorted numerically by used storage total.
> 
> Does not seem to work. Is there any update (on requirements/implementation) that I am missing?

This would depend on the project table re-design, whose exact implementation is not yet known https://community.openproject.org/work_packages/18800 (and is still missing the table style et al.)

oliverguenther commented:
Concerning the missing repository count. Are you running delayed jobs ? 

If you were not running it, make sure to clear the rails cache or reboot the server to update the count after you started the job runner.

oliverguenther committed:
Pin acts_as_countable to 2471265.

acts_as_countable has no release yet, and the repository is managed by
us, this commit thus pins the repository at this specific ref instead.

NobodysNightmare commented:
> Are you running delayed jobs ?

Jup... My database is also displaying the latest info correctly...

oliverguenther commented:
I seem to have found a case where the `root_url` is not properly set and thus no storage is retrieved at all. However, if the storage was updated at least once, this issue should not causing that behavior. I'll see   If I can reproduce it.

NobodysNightmare commented on app/models/project.rb:
```diff
@@ -929,6 +932,19 @@ def archive!
 
+  def required_storage
+    Rails.cache.fetch("project##{id}/project_storage", expires_in: 1.hour) do
+      storage, = count_for(:required_project_storage)
```
I really thought to see a typo here. But I believe you are intentionally putting that `,` there...

I did not yet find the advantage over `count_for(...).first`, which is what I would propose ^^

oliverguenther replied:
Heh, matter of preference. But I agree your proposal is more explicit, and I like that.

oliverguenther committed:
Use more explicit partial parameter extraction

oliverguenther committed:
Force setting of root_url for local git repositories

When the root_url is not delivered explicitly, the scm adapter
is missing it even though it may be able to count it, e.g.,
when adding existing, local repositories

NobodysNightmare commented:
I can probably de-escalate the alarm level.

You are using the rails cache. I believe the cache was just filled with zero values before it got the right numbers. Will call back when the problem is not solved after 1 hour or a cache invalidation.

edit: As expected, the problem solved itself over time ^^

NobodysNightmare commented:
``` ruby
pry(main)> Project.find(1).count_for(:required_project_storage).first
=> {"attributes.attachments"=>1766242, :label_repository=>452600, :total=>2218842}
```

Note how the first key is a string and the other keys are symbols... Looks unintentional, but I did not check the code yet.

oliverguenther commented:
Yeah, thats simply the way the labels are defined in the `acts_as_countable` uses (attachment and repository model). As they're meant to be input to `I18n.t`, I didn't really care .. it doesn't look very pretty though :flushed:

NobodysNightmare commented on app/models/repository.rb:
```diff
@@ -165,6 +175,18 @@ def relative_path(path)
+  def required_disk_storage
+    if scm.storage_available?
+      if storage_updated_at.nil? ||
+         storage_updated_at < 12.hours.ago
```
Shouldn't such a timeout be configurable via `settings.yml`?

oliverguenther replied:
Yeah, you're goddamn right.

NobodysNightmare commented on app/models/repository.rb:
```diff
@@ -165,6 +175,18 @@ def relative_path(path)
+        Delayed::Job.enqueue ::Scm::StorageUpdaterJob.new(self)
+      end
+
+      required_storage_bytes
```
Just a note so that we know it, not neccessarily a request for change:

This seems to be the reason why I will see 0 Bytes for the next hour:
1. ask rails cache if it knows required space
2. calculate required space
3. enqueue a delayed job for repo size
4. return currently known repo size (on init this is 0)
5. the delayed job will update the known repo size

Upon requesting this value I will _never_ get the latest value, because with 99.99% certainty I get the cached result before a new one is calculated in the delayed job.

oliverguenther replied:
That is exactly right and the reason this value may be externally updated (through the Sys api) and we may have to update this value regularly. I chose not to do this simply because the exact requirement on timeliness wasn't clear, and  in the current state, all those possibilities remain easy to implement.

NobodysNightmare commented on app/models/project.rb:
```diff
@@ -929,6 +932,18 @@ def archive!
   end
 
+  def required_storage
+    Rails.cache.fetch("project##{id}/project_storage", expires_in: 1.hour) do
```
This timeout is also a candidate for a `Setting`

NobodysNightmare commented on app/views/admin/projects.html.erb:
```diff
@@ -40,6 +40,29 @@ See doc/COPYRIGHT.rdoc for more details.
+      <dt class="attributes-key-value--key">Count</dt>
+      <dd class="attributes-key-value--value-container">
+          <div class="attributes-key-value--value -text">
+              <span title="Count"><%= Project.count %> projects</span>
```
Help! Localize us!

NobodysNightmare commented on app/views/storage/_required_storage.html.erb:
```diff
@@ -0,0 +1,41 @@
+  <% (storage.keys - [:total]).each do |type| %>
+    <% if storage[type] > 0 %>
+    <%= l(type) %>: <%= number_to_human_size(storage[type], precision: 2) %>
+    &nbsp;
```
`&nbsp;` is something that rings an alarm for me...

Isn't there CSS that should control these things?

oliverguenther replied:
As a tooltip content may receive no styling and I cannot output HTML in there (I don't know why exactly that is), I chose to separate entries in that tooltip them by a forced whitespace. I'm not particular fond of the way the tooltip looks like, though and am open for ideas.

NobodysNightmare replied:
If there only was a way to display native OpenProject tooltips: https://github.com/opf/openproject/pull/3246 :smile: 

Probably this will improve our situation once it is fixed... I am not sure of a proper solution either, sadly...

oliverguenther replied:
Ah, gems hidden inside a giant stuck PR.. awesome! :+1:

NobodysNightmare commented on config/routes.rb:
```diff
@@ -520,6 +520,7 @@
 
   scope controller: 'sys' do
     match '/sys/projects.:format', action: 'projects', via: :get
+    match '/sys/projects/:id/repository/storage', action: 'update_required_storage', via: :get
```
Only talking about naming here, but since this is an endpoint that does not return anything and is only meant to be pushed, I would vote for something with a verb, e.g. `update_storage`, rather than `storage`

oliverguenther replied:
:+1:

oliverguenther committed:
Configurable cache expiry

This commit:

* Adds cache expiry time in minutes for projects and repositories
* renames the sys route to clarify it does not return anything
* consolidate countable labels as strings

NobodysNightmare commented:
This is a little late now, but thinking about it I am not quite sure whether your settings should go into the `settings.yml` or the `configuration.yml`.

Not sure what the difference between both is, but most `settings` I can come up with are configurable via the Web UI... However, this is probably not a good indicator of where something should go ^^

oliverguenther commented:
Oh, okay. I did not really know of that virtual separation. I don't really mind where it goes.

NobodysNightmare commented:
Me neither and I personally put everything into settings.yml

Just not sure whether there is a rationale when to use which one.

@linki do you know more? Think you introduced the config...

oliverguenther committed:
Correct localization of project view

linki commented:
i didn't introduce that :)

but i think of them in the following way: `OpenProject::Configuration` are system-wide boot-time configuration options that are not changeable at runtime. they can only be changed with a rails-server reboot. `Setting` are instance-wide, runtime-changeable configuration options. they persist, though, so they need a database. since they need a database they cannot (shouldn't) be used at rails-boot time (e.g. in initializers). note that this is currently not true, therefore we need a database for rake tasks like precompilation. if you depend on a setting in sth. like precompilation you might not be able to build them deterministically.

the difference gets more clear in the multitenancy scope. there, settings are scoped to a single instance, while configuration is shared by all instances (not sure if that's desirable, but that's how it is). in the context of repositories that could mean that the set of available repository adapters could be a configuration (and triggers loading of the several required rubygems at boot-time) whereas the individual repository options would need to be settings (because they are scoped to an instance and added at runtime i presume)

oliverguenther committed:
Fix invalid expectactions in specs due to refactoring

oliverguenther commented:
@linki Thanks for the clarification. I guess that in this case, it could go both ways. I can foresee that these values should become configurable through the frontend, and thus I'd reason there to be no harm in putting them into settings right now. They do not relate to anything at startup.

@NobodysNightmare just FYI: Specs were failing due to refactored `storage,  = foo`, which was fine when my expects simply returned nil. If you agree with the settings choice, then I think we're done if travis agrees.

NobodysNightmare commented:
@oliverguenther Did anyone ever have a look at `acts_as_countable`? Quite clever of you to hide it inside its own repository ;-)

Just kidding. LGTM if travis agrees! :+1:

oliverguenther commented:
I lay all blame on @ulferts , I was inlining it to hell! :smiling_imp:
